src/overlords/offense/roomPoisoner.ts

```diff
@@ -0,0 +1,159 @@
+    	this.sourcesWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+                                                                    s => s.structureType == STRUCTURE_WALL &&
+                                                                    !s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+    }
+
+	refresh() {
+		super.refresh();
+		this.controllerWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+            s => s.structureType == STRUCTURE_WALL &&
+            s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+        this.sourcesWallSites = (this.room && this.room.controller) ? _.filter(this.room.constructionSites,
+            s => s.structureType == STRUCTURE_WALL &&
+            !s.pos.isNearTo(this.room!.controller!.pos)) : undefined;
+	}
+
+	init() {
+		let isSafe = this.room && !this.room.dangerousPlayerHostiles.length;
+		let isNotReservedByEnemy = !(this.room && this.room.controller && this.room.controller.reservation && this.room.controller.reservation.ticksToEnd > 500);
+
+		//spawn required creeps to contaminate if visible + safe + notRserved + notPoisoned, else spawn reserved is reserved by enemy
+		
+		if(this.pos.isVisible && isSafe && !this.initializer.memory.isPoisoned){
+			if(isNotReservedByEnemy){
+				if(!(this.room && this.room.my)) this.wishlist(1, Setups.infestors.claim);
+				this.wishlist(1, Setups.roomPoisoner);
+			}else{
+				if(!(this.room && this.room.my)) this.wishlist(1, Setups.infestors.controllerAttacker);	
+			}
+		}
+	}
+
+	private findStructureBlockingController(roomPoisoner: Zerg): Structure | undefined {
+		const blockingPos = Pathing.findBlockingPos(roomPoisoner.pos, roomPoisoner.room.controller!.pos,
+													_.filter(roomPoisoner.room.structures, s => !s.isWalkable));
+		if (blockingPos) {
+			const structure = blockingPos.lookFor(LOOK_STRUCTURES)[0];
+			if (structure) {
+				return structure;
+			} else {
+				log.error(`${this.print}: no structure at blocking pos ${blockingPos.print}! (Why?)`);
+			}
+		}
+	}
+
+	private handleRoomPoisoner(roomPoisoner: Zerg): void {
+		//always recharge if energy == 0 even in spawn room.
+		if (roomPoisoner.carry.energy == 0) {
+			roomPoisoner.task = Tasks.recharge();
+			return;
+		}
+		// Ensure you are in the assigned room
+		if (roomPoisoner.room == this.room && !roomPoisoner.pos.isEdge) {
+			//corner case: unclaimed controller blocked, while sources not 100% blocked
+			if(!this.room.my && this.sourcesWallSites && this.controllerWallSites &&
+				this.controllerWallSites.length == 0 &&  this.sourcesWallSites.length > 0){
+				
+				const dismantleTarget = this.findStructureBlockingController(roomPoisoner);
+				if (dismantleTarget) {
+					roomPoisoner.task = Tasks.dismantle(dismantleTarget);
+					return;
+				}	
+			}
+
+			// upgrade controller to level 2 to unlock walls			
+			if (this.room && this.room.controller && this.room.my &&
```

Comment:
this type of cascaded task assignment might be better broken up into a bunch of different private methods like the way I have it organized in `WorkerOverlord`